catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/ProjectUploadRatingDialogTest.java

```diff
@@ -0,0 +1,172 @@
+import static org.catrobat.catroid.ui.ProjectUploadActivity.NUMBER_OF_UPLOADED_PROJECTS;
+import static org.catrobat.catroid.ui.ProjectUploadActivity.PROJECT_DIR;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@RunWith(AndroidJUnit4.class)
+public class ProjectUploadRatingDialogTest {
+
+	private ProjectUploadController projectUploadController;
+	private Bundle bundle;
+	private SharedPreferences sharedPreferences;
+	private static final String PROJECT_NAME = "projectName";
+
+	@Rule
+	public BaseActivityTestRule<ProjectUploadTestActivity> activityTestRule =
+			new BaseActivityTestRule<>(ProjectUploadTestActivity.class, false, false);
+
+	@Before
+	public void setUp() throws Exception {
+		Context context = InstrumentationRegistry.getTargetContext();
+		sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+		bundle = new Bundle();
+
+		Project project = new DefaultProjectCreator().createDefaultProject(PROJECT_NAME,
+				InstrumentationRegistry.getTargetContext(),	false);
+
+		Sprite firstSprite = new SingleSprite("firstSprite");
+		project.getDefaultScene().addSprite(firstSprite);
+		ProjectSaveTask.task(project, context);
+
+		Intent intent = new Intent();
+		intent.putExtra(PROJECT_DIR, project.getDirectory());
+		activityTestRule.launchActivity(intent);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+	}
+
+	@Category({Cat.AppUi.class, Level.Smoke.class})
+	@Test
+	public void testUploadControllerGetsCalled() {
+		uploadProject();
+
+		projectUploadController = activityTestRule.getActivity().getProjectUploadController();
+
+		verify(projectUploadController, times(1)).startUpload();
+	}
+
+	@Category({Cat.AppUi.class, Level.Smoke.class})
+	@Test
+	public void uploadSuccessRatingDialogShowing() {
+		uploadProject();
+
+		InstrumentationRegistry.getInstrumentation().runOnMainSync(() -> {
+			activityTestRule.getActivity().showUploadDialog();
+			activityTestRule.getActivity().onReceiveResult(UPLOAD_RESULT_RECEIVER_RESULT_CODE, bundle);
+		});
+
+		onView(withText(R.string.rating_dialog_rate_now))
+				.check(matches(isDisplayed()));
+	}
+
+	@Category({Cat.AppUi.class, Level.Smoke.class})
+	@Test
+	public void uploadFailRatingDialogNotShowing(){
+		uploadProject();
+
+		InstrumentationRegistry.getInstrumentation().runOnMainSync(() -> {
+			activityTestRule.getActivity().showUploadDialog();
+			activityTestRule.getActivity().onReceiveResult(0,bundle);
+		});
+
+		onView(withText(R.string.rating_dialog_rate_now))
+				.check(doesNotExist());
+	}
+
+	public void uploadProject(){
+		sharedPreferences.edit()
+				.putInt(NUMBER_OF_UPLOADED_PROJECTS, 1)
+				.commit();
+
+		onView(withId(R.id.next))
+				.perform(click());
+
+		onView(withText(R.string.next))
+				.perform(click());
+	}
+
+
+	public static class ProjectUploadTestActivity extends ProjectUploadActivity {
```

Comment:
please dont dont do test doubles by overriding the object under test.

theres other ways around this, one would be a simple setter method that is @visibleForTesting to replace the real controller with a mock.

Reply:
ok, i did some research and i take this back. it is pretty much impossible to mock anything from an activity before its onCreate is called. this just sucks and we probably have to live with it.